bin/interactive.py

```diff
@@ -298,9 +289,9 @@ def kill_submission_handler(signal, frame):
             submission_status = status
 
             # Close the output stream.
-            os.close(team_out)
+            validator.stdin.close()
```

Comment:
In fact, it could be that there is still some data being buffered in the submission->validator pipe, and I would think that closing the read-end of that would destroy the data, while closing the write-end still allows the validator to read to the end.

Reply:
Yes, and it turns out that it is the same pipe, just from the other end :wink:

Reply:
Also note that `validator.stdin` is passed to either `team_tee` or `submission`, depending on whether `interaction == True`. I changed it to close "on the other end" to avoid the conditional :slightly_smiling_face:

Reply:
Does it make a difference from which end a pipe is closed? ^^'

Reply:
You want to close the write end to signal to the child reading on the other end that no more data is coming. Closing the read end possibly doesn't even do anything since the child has it's own copy of the file descriptor.

Reply:
https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stdin

> If the _stdin_ argument was `PIPE`, this attribute is a writeable stream object as returned by `open()`. [...] If the _stdin_ argument was not `PIPE`, this attribute is `None`.

So, I tried to close the stream from "the other end", but that doesn't work, because `submission.stdout is None` :joy: I'm assuming that `subprocess.Popen` simplifies from the fact that the file descriptor is duplicated like you mentioned, so that you always have a "single point of contact" to manage the stream?

Reply:
Right, `validator.stdin` indeed gives the `write` end of the stream for the `parent`, not the `read` end of the stream used by the `child`. I suppose `subprocess` is quite a bit higher level than the kernel functions :D